packages/swing-store-simple/src/simpleSwingStore.js

```diff
@@ -232,49 +229,142 @@ function makeSwingStore(dirPath, forceReset = false) {
+    );
+    insistStreamPosition(startPosition);
+    insistStreamPosition(endPosition);
+    assert(startPosition.itemCount <= endPosition.itemCount);
+
+    if (endPosition.itemCount === 0) {
+      return [];
+    } else {
+      const readStatus = `read-${statusCounter}`;
+      statusCounter += 1;
+      streamStatus.set(streamName, readStatus);
+      stream.length = endPosition.itemCount;
+      let pos = startPosition.itemCount;
+      function* reader() {
+        while (pos < stream.length) {
+          const statusInner = streamStatus.get(streamName);
+          assert(
+            statusInner === readStatus,
+            X`can't read stream ${q(streamName)}, it's been closed`,
+          );
+          const result = stream[pos];
+          pos += 1;
+          yield result;
+        }
+        const statusEnd = streamStatus.get(streamName);
+        assert(
+          statusEnd === readStatus,
+          X`can't read stream ${q(streamName)}, it's been closed`,
+        );
+        streamStatus.delete(streamName);
       }
+      return reader();
     }
   }
 
   /**
-   * Append an item to a stream.
+   * Obtain a writer for a stream.
    *
-   * @param {string} streamName  The stream to append to
-   * @param {string} item  The item to append to it
+   * @param {string} streamName  The stream to be written
    */
-  function appendItem(streamName, item) {
-    assert.typeof(
-      streamName,
-      'string',
-      X`non-string stream name ${streamName}`,
-    );
-    assert(streamName.match(/^[-\w]+$/));
-    let stream = streams.get(streamName);
-    if (!stream) {
-      stream = [];
-      streams.set(streamName, stream);
+  function openWriteStream(streamName) {
+    insistStreamName(streamName);
+    let streamTemp = streams.get(streamName);
+    if (!streamTemp) {
+      streamTemp = [];
+      streams.set(streamName, streamTemp);
+      streamStatus.set(streamName, 'write');
+    } else {
+      const status = streamStatus.get(streamName);
+      assert(
+        !status,
+        X`can't write stream ${q(streamName)} because it's already in use`,
+      );
+    }
+    const stream = streamTemp;
+    const writeStatus = `write-${statusCounter}`;
+    statusCounter += 1;
+    streamStatus.set(streamName, writeStatus);
+
+    /**
+     * Write to a stream.
+     *
+     * @param {string} item  The item to write
+     * @param {Object|null} position  The position to write the item
+     *
+     * @returns {Object} the new position after writing
+     */
+    function write(item, position) {
+      const status = streamStatus.get(streamName);
+      assert(
+        status === writeStatus,
+        X`can't write to closed stream ${q(streamName)}`,
+      );
+
+      if (position) {
+        insistStreamPosition(position);
+      } else {
+        position = STREAM_START;
+      }
+      stream[position.itemCount] = item;
+      return { itemCount: position.itemCount + 1 };
```

Comment:
let's harden this one too

Reply:
Yeah, I just basically put in a harden anywhere there was a `return {`...